synapse/handlers/auth.py

```diff
@@ -772,13 +774,114 @@ def _auth_dict_for_flows(
             "params": params,
         }
 
+    async def refresh_token(
+        self,
+        refresh_token: str,
+        valid_until_ms: Optional[int],
+    ) -> Tuple[str, str]:
+        """
+        Consumes a refresh token and generate both a new access token and a new refresh token from it.
+
+        The consumed refresh token is considered invalid after the first use of the new access token or the new refresh token.
+
+        Args:
+            refresh_token: The token to consume.
+            valid_until_ms: The expiration timestamp of the new access token.
+
+        Returns:
+            A tuple containing the new access token and refresh token
+        """
+
+        # Verify the token signature first before looking up the token
+        if not self.verify_refresh_token(refresh_token):
+            raise SynapseError(400, "invalid refresh token")
+
+        existing_token = await self.store.lookup_refresh_token(refresh_token)
+        if existing_token is None:
+            raise SynapseError(400, "refresh token does not exist")
+
+        if (
+            existing_token.has_next_access_token_been_used
+            or existing_token.has_next_refresh_token_been_refreshed
+        ):
+            raise SynapseError(400, "refresh token isn't valid anymore")
```

Comment:
should these be 403s, maybe?

It'd also be nice to return something other than `M_UNKNOWN` as the error code. (Error behaviour should be covered in the MSC.)